src/simsopt/geo/qfmsurface.py

```diff
@@ -0,0 +1,173 @@
+from scipy.optimize import minimize, least_squares, NonlinearConstraint
+import numpy as np
+from simsopt.geo.surfaceobjectives import QfmResidual
+
+
+class QfmSurface():
+
+    def __init__(self, biotsavart, surface, label, targetlabel):
+        self.bs = biotsavart
+        self.surface = surface
+        self.label = label
+        self.targetlabel = targetlabel
+
+    def qfm_label_constraint(self, x, derivatives=0):
+        """
+        Returns the residual
+
+        0.5 * (label - labeltarget)^2
+
+        and optionally the gradient wrt surface params.
+        """
+        assert derivatives in [0, 1]
+
+        s = self.surface
+        sdofs = x
+        nsurfdofs = x.size
+        bs = self.bs
+
+        s.set_dofs(sdofs)
+        l = self.label.J()
+        rl = l-self.targetlabel
+        val = 0.5 * rl ** 2
+
+        if derivatives:
+            dl = self.label.dJ_by_dsurfacecoefficients()
+            dval = rl*dl
+            return val, dval
+        else:
+            return val
+
+    def qfm_objective(self, x, derivatives=0):
+        """
+        Returns the residual
+
+        f(x)
+
+        and optionally the gradient wrt surface params where f(x) is the QFM
+        residual defined in surfaceobjectives.py.
+        """
+        assert derivatives in [0, 1]
+
+        s = self.surface
+        sdofs = x
+        nsurfdofs = x.size
+        bs = self.bs
+
+        s.set_dofs(sdofs)
+        qfm = QfmResidual(s, bs)
+
+        r = qfm.J()
+
+        if derivatives == 0:
+            return r
+        else:
+            dr = qfm.dJ_by_dsurfacecoefficients()
+            return r, dr
+
+    def qfm_penalty_constraints(self, x, derivatives=0, constraint_weight=1.):
+        """
+        Returns the residual
+
+        f(x) + 0.5 * constraint_weight * (label - labeltarget)^2
+
+        and optionally the gradient and Hessian wrt surface params where f(x)
+        is the QFM residual defined in surfaceobjectives.py.
+        """
+
+        assert derivatives in [0, 1, 2]
+        s = self.surface
+        sdofs = x
+        nsurfdofs = x.size
+        bs = self.bs
+
+        s.set_dofs(sdofs)
+
+        qfm = QfmResidual(s, bs)
+
+        r = qfm.J()
+
+        l = self.label.J()
+        rl = l-self.targetlabel
+
+        val = r + 0.5 * constraint_weight * rl**2
+        if derivatives == 0:
+            return val
+
+        dr = qfm.dJ_by_dsurfacecoefficients()
+
+        dl = self.label.dJ_by_dsurfacecoefficients()
+
+        dval = dr + constraint_weight*rl*dl
+        if derivatives == 1:
+            return val, dval
+
+        d2r = qfm.d2J_by_dsurfacecoefficientsdsurfacecoefficients()
+        d2l = self.label.d2J_by_dsurfacecoefficientsdsurfacecoefficients()
+        d2val = d2r  \
+            + constraint_weight*rl*d2l + constraint_weight*dl[:, None]*dl[None, :]
+
+        return val, dval, d2val
+
```

Comment:
Can we have another function `minimize_qfm_penalty_constraints` with LBFGS and SLSQP supplied as choices to an argument called method

```
def minimize_qfm_penalty_constraints(self, tol=1e-3, maxiter=1000,
                                               constraint_weight=1, method=<default of LBFGS/SLSQP>)
```

Reply:
I added the docstring. 

I am not sure what you want to achieve with minimize_qfm_penalty_constraints. Right now, minimize_qfm_exact_constraints_SLSQP solves a constrained problem while minimize_qfm_penalty_constraints_LBFGS solves an unconstrained penalized problem. For example, constraint_weight is not used for minimize_qfm_exact_constraints_SLSQP.